### ISSUE TITLE
app returning state or destroy(kill) func?

### ISSUE THREAD

**Issue by sergey-shpak** [jorgebucaran/hyperapp#863]
Hyperapp#2 app should return something useful to dev userland, several options:

1. **emit app state changes**, can be useful when you don't have any possibility to add subscription directly to the app: custom elements, etc 
```javascript
const onStateChange = app(props)
onStateChange(state => {
  // do your external state magic
})
``` 

2. **return kill function**. After 'app' purity discussion, we came that modifying dom is a side effect, and should be implemented as internal subscription, as result - 'kill/destroy' function is a trigger to unsubscribe all the app subscriptions
```javascript
const kill = app(props)
onExternalChange(() => kill()) // pass new state/reason as arg?
```

For performance and memory reasons I would go with 2, but 1 can be quite useful too, any suggestions?

**Comment by sergey-shpak** [comment]
2 can be easy as
```javascript 
var unsubscribe = function(){ patchSubs(subs, [], dispatch)  }
return function kill(s){ dispatch([ s || state, [ unsubscribe ] ]) }
```
@jorgebucaran

**Comment by sergey-shpak** [comment]
as another suggestion, 'kill' can be propagated through the app action to properly shutdown if needed, something like:
```javascript
// ... app return
function unsubscribe(state){
  return [state, [ function(){ patchSubs(subs, [], dispatch = function(){} ) } ]]
}
return function kill(){
  return dispatch(props.kill ? [props.kill, unsubscribe] : unsubscribe)
}

// and then
app({ 
  init, view, subscriptions, 
  kill: (state, resolve) => 
    [{ ...state, closing: true }, someEffect({ action: resolve }) ] 
})
```
* 'kill' is optional app prop
* 'kill' can be modified to propagate/dispatch reason/state or even last app action

**Comment by jorgebucaran** [comment]
@sergey-shpak That's very clever. `kill` as an app prop can be a bit unclear, though. _If_ this API was implemented, it'd be better to just return the function.

**Comment by sergey-shpak** [comment]
@jorgebucaran thanks, after some thoughts I couldn't imagine real need to pass state to `kill` app, so you are right and implementation for 2 stay simple
```javascript
function unsubscribe(){ patchSubs(subs, [], dispatch = function(){} ) }
return function kill(){ return dispatch(function destroy(state){ return [state, [ unsubscribe ]] }) }
```
* 'destroy' action is used especially to notify devtools (do not to kill app silently)
* this will work only if 'dom render' is moved into subscriptions

**Comment by sergey-shpak** [comment]
or even shorter
```javascript
function unsubscribe(){ patchSubs(subs, [], dispatch = function(){} ) }
return function kill(s){ return s === state ? [s, [ unsubscribe ]] : dispatch(kill) }
```
and then self-destroying becomes possible
```jsx
const kill = app({
  view: state => <button onclick={ kill } >kill</button>
})
```
:rocket: :rocket: :rocket:

**Comment by jorgebucaran** [comment]
@sergey-shpak This app can kill itself. Interesting.

Should `kill` be responsible for clearing the DOM too? What else should it do?

Or what about a higher order app (function) that can cancel all subscriptions?

**Comment by sergey-shpak** [comment]
@jorgebucaran, since dom manipulations (vdom render) should be wrapped into subscription (do not to create side effect) it should clear itself on unsubscribe called (which is what 'kill' exactly does ;) )

```javascript
// somewhere inside of hyperapp#2
function render(dispatch, props){
  return function unrender(){
    // restore/clear dom, props.node, etc
  }
}
// this is hacky a little, but it is the only way 
// how to pass state to sub without sub reinit 
function vdom(props){
  // deferred dom patch with props.node, props.vdom
  return [render, { props.node }]
}

// somewhere inside app function
export function app(props){
// ...
function subscriptions(state){
  return [].concat(props.subscriptions(state), vdom({ node, view, state }) )
}
// ...
}
```
when 'kill' is called - 'render' sub is unsubscribed and 'unrender' is called

**Comment by sergey-shpak** [comment]
@jorgebucaran HOF for app 'kill' is too dirty (you don't have explicit access to 'patchSubs' from external, you need to trigger subscriptions update and then return empty array, fake subs to catch dispatch, etc)

**Comment by sergey-shpak** [comment]
on the other hand, sometimes emitting app state into external world is needed, for example to sync storage or notify outer app, currently 'subscriptions' can be used to achieve this:
```javascript
const storageSub = () => {
  return function unsubscribe(){}
}
const storage = (state) => {
  // ... saving each state change to local storage
  return [storageSub, {}]
}
app({ subscriptions: state => [ storage(state) ]})
```
this will work, but it has some hacky smell:
1. 'subscriptions' designed to work in opposite way
2. it's mimic subscription to emit state changes

As result, it is one more argument to implement *emit app state* API (either as 'app return' or as standalone API)

**Comment by jorgebucaran** [comment]
@sergey-shpak I prefer your (2) suggestion. Rather than `onStateChange` or even using a subscription, I suggest using middleware to synchronize your state to local storage.

**Comment by sergey-shpak** [comment]
@jorgebucaran ok good point, so we stay with 2, PR now or after dom manupulations (render) are moved to subscription?

**Comment by mrozbarry** [comment]
So what's the main point of returning either of these options? Is there a real use case for distributing the Hyperapp state?

And also,

> After 'app' purity discussion, we came that modifying dom is a side effect, and should be implemented as internal subscription, as result - 'kill/destroy' function is a trigger to unsubscribe all the app subscriptions

This seems wrong. Modifying the DOM is certainly a side-effect, but is it an internal subscription? My understanding of subscriptions is they are long-run side-effect driven, and can be stateful, but for the DOM re-rendering, the entire state is required, which will cause frequent subscription restarts, which will lose the state. If you store the vdom state in a global/module variable to counter-act the restarts, then it would be just as easy to implement as a side-effect.

I'm not even sure why this relates to an app kill switch. Implementing a kill switch into an app is trivial, and cleaning up the dom should also be trivial, since the user knows which DOM element is used as the root node.

Sure, a kill switch could be a really nice HoF/HoA, but I don't see any reason why that needs to be in core. Not until we figure out if it's something that >50% of apps need.

**Comment by sergey-shpak** [comment]
@mrozbarry real cases are
- **emitting app state**: sync storage, emit state to outer app
- **killing app**: custom elements, composing apps

Hyperapp#2 provides api for long lived side-effects - subscriptions, so if dom manipulation is a side effect it should be treated as a subscription. But not to complicate dev user land (and to fit current app API) the subscription should be added internally. ([example without sub restarting](https://github.com/jorgebucaran/hyperapp/issues/863#issuecomment-515737153))

Dom manipulation(render) as subscription is nice not only because it makes hyperapp internally more decoupled and following purity principals, but because perfectly fits into app 'kill' functionality (at the end of the day app 'kill' has to unsubscribe all the app subscriptions, avoiding app memory leaks, and other side effects)

Hyperapp#2 core contains functionality which cannot be implemented any other way (even if less than 10% apps need it). I cannot imagine any 'normal' implementation of app 'kill' with HoF/HoA

**Comment by sergey-shpak** [comment]
@mrozbarry moreover, I could be wrong but destroying root dom node from hof/hoa won't actually delete it because it will be referenced from internal hyperapp#2 'node' variable (meaning dom will be visually cleared but memory won't be cleared -> memory leak)

**Comment by jorgebucaran** [comment]
> @mrozbarry Not until we figure out if it's something that >50% of apps need.

Agreed.

**Comment by mrozbarry** [comment]
> Hyperapp#2 provides api for long lived side-effects - subscriptions, so if dom manipulation is a side effect it should be treated as a subscription. But not to complicate dev user land (and to fit current app API) the subscription should be added internally. (example without sub restarting)

@sergey-shpak that example will restart if `state` changes, and `state` must change for a re-render, so I have no idea how that would work without restarting the sub each time.

> moreover, I could be wrong but destroying root dom node from hof/hoa won't actually delete it because it will be referenced from internal hyperapp#2 'node' variable

I do agree with the memory leak issue, but I think we'd have to address that if we need an internal kill switch implemented. I did some looking, and at least both Vue and React support destroy/unmount features, respectively, and I'm guessing they do any memory cleanup necessary. I think we might need to ask the community how many apps need to be unmounted, and how frequently to really know how necessary it is.

**Comment by zaceno** [comment]
One case for the kill-command is when using hyperapp to implement custom elements. When a custom element is removed, we need to ensure any subscriptions it started are shut off. However, that’s fully doable in userland as long as dispatch is available from the outside somehow (like with the current enhancer/middleware feature). So I don’t think hyperapp necessarily needs a built in kill switch, unless we get rid of the middleware feature.

**Comment by jorgebucaran** [comment]
@sergey-shpak @zaceno @mrozbarry Unless the current middleware feature approach breaks down or turns out insufficient, I'm going with a `kill` function as the app's return value. 

Sergey: If you want to submit a PR, go ahead. 👋:)

**Comment by sergey-shpak** [comment]
> @sergey-shpak that example will restart if `state` changes, and `state` must change for a re-render, so I have no idea how that would work without restarting the sub each time.

@mrozbarry if you look closer to the [provided example](https://github.com/jorgebucaran/hyperapp/issues/863#issuecomment-515737153) 'vdom' function returns the same subscription each time without passing state to it `[render, { props.node }]`, as result it won't be restarted on each state change. (it is a little tricky, because side effect is split into two functions - first to catch and render vdom with state, and the second to return unsubscribe function)

@zaceno :+1:  'kill' is needed not only for custom elements, but also for any widget or multiple app composition. 

@zaceno @jorgebucaran @mrozbarry 
We already found that `onStateChange` can be implemented with HoF so it is clear now, 
but I can't imagine any proper solution based on middleware/hof/hoa that can properly kill app (destroy internal dom node reference and unsubscribe all the subscriptions), if you show me a real example I'm 100% will use it and will forget about 'kill' function.

*apps amount requiring 'kill' fn could be even close to 0%, but this could mean that apps are not created with ha#2 because it doesn't have proper kill fn, you never know for sure )

**Comment by mrozbarry** [comment]
> if you look closer to the provided example 'vdom' function returns the same subscription each time without passing state to it [render, { props.node }], as result it won't be restarted on each state change. (it is a little tricky, because side effect is split into two functions - first to catch and render vdom with state, and the second to return unsubscribe function)

I'm going to have to see some real implemented and tested code. This looks like it does a single deferred render perfectly fine, but I'm not convinced of the persistence of the sub without it restarting if you pass state as a prop to it.

**Comment by sergey-shpak** [comment]
@mrozbarry @jorgebucaran 
real working example with vdom render as subscription (as proposed this should be done internally not in userland) https://codepen.io/sergey-shpak/pen/MNpWdj
